tsdb/engine/tsm1/tombstone.go

```diff
@@ -154,94 +183,67 @@ func (t *Tombstoner) writeTombstone(tombstones []Tombstone) error {
 }
 
 func (t *Tombstoner) readTombstone() ([]Tombstone, error) {
-	f, err := os.Open(t.tombstonePath())
-	if os.IsNotExist(err) {
-		return nil, nil
-	} else if err != nil {
-		return nil, err
-	}
-	defer f.Close()
-
-	var b [4]byte
-	_, err = f.Read(b[:])
-	if err != nil {
-		// Might be a zero length file which should not exist, but
-		// an old bug allowed them to occur.  Treat it as an empty
-		// v1 tombstone file so we don't abort loading the TSM file.
-		return t.readTombstoneV1(f)
-	}
+	var tombstones []Tombstone
 
-	if _, err := f.Seek(0, os.SEEK_SET); err != nil {
+	if err := t.Walk(func(t Tombstone) error {
+		tombstones = append(tombstones, t)
+		return nil
+	}); err != nil {
 		return nil, err
 	}
-
-	if binary.BigEndian.Uint32(b[:]) == v2header {
-		return t.readTombstoneV2(f)
-	}
-	return t.readTombstoneV1(f)
+	return tombstones, nil
 }
 
 // readTombstoneV1 reads the first version of tombstone files that were not
 // capable of storing a min and max time for a key.  This is used for backwards
 // compatibility with versions prior to 0.13.  This format is a simple newline
 // separated text file.
-func (t *Tombstoner) readTombstoneV1(f *os.File) ([]Tombstone, error) {
-	var b []byte
-	var err error
-	b, err = ioutil.ReadAll(f)
-	if err != nil {
-		return nil, err
-	}
-
-	lines := strings.TrimSpace(string(b))
-	if lines == "" {
-		return nil, nil
-	}
-
-	tombstones := []Tombstone{}
-
-	for _, line := range strings.Split(string(b), "\n") {
+func (t *Tombstoner) readTombstoneV1(f *os.File, fn func(t Tombstone) error) error {
+	r := bufio.NewScanner(f)
+	for r.Scan() {
+		line := r.Text()
 		if line == "" {
 			continue
 		}
-		tombstones = append(tombstones, Tombstone{
+		if err := fn(Tombstone{
 			Key: line,
 			Min: math.MinInt64,
 			Max: math.MaxInt64,
-		})
+		}); err != nil {
+			return err
+		}
 	}
-	return tombstones, nil
+	return r.Err()
 }
 
 // readTombstoneV2 reads the second version of tombstone files that are capable
 // of storing keys and the range of time for the key that points were deleted. This
 // format is binary.
-func (t *Tombstoner) readTombstoneV2(f *os.File) ([]Tombstone, error) {
+func (t *Tombstoner) readTombstoneV2(f *os.File, fn func(t Tombstone) error) error {
 	// Skip header, already checked earlier
 	if _, err := f.Seek(4, os.SEEK_SET); err != nil {
-		return nil, err
+		return err
```

Comment:
nit: I know it's not in your change, but maybe we could have a `v2HeaderSize` or something? Could be referenced in the byte slice in `(t *Tombstoner) Walk` too.